contrib/database/sql/conn.go

```diff
@@ -167,6 +211,48 @@ func WithSpanTags(ctx context.Context, tags map[string]string) context.Context {
 	return context.WithValue(ctx, spanTagsKey, tags)
 }
 
+// tryStartTrace will create a span using the given arguments, but will act as a no-op when err is driver.ErrSkip.
+func (tp *traceParams) tryStartTrace(ctx context.Context, qtype queryType, query string, startTime time.Time, sqlCommentCarrier *tracer.SQLCommentCarrier) (span tracer.Span) {
+	if _, exists := tracer.SpanFromContext(ctx); tp.cfg.childSpansOnly && !exists {
+		return nil
+	}
+	name := fmt.Sprintf("%s.query", tp.driverName)
+	opts := []ddtrace.StartSpanOption{
+		tracer.ServiceName(tp.cfg.serviceName),
+		tracer.SpanType(ext.SpanTypeSQL),
+		tracer.StartTime(startTime),
+	}
+	if !math.IsNaN(tp.cfg.analyticsRate) {
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, tp.cfg.analyticsRate))
+	}
+	span, _ = tracer.StartSpanFromContext(ctx, name, opts...)
+	resource := string(qtype)
+	if query != "" {
+		resource = query
+	}
+	span.SetTag("sql.query_type", string(qtype))
+	span.SetTag(ext.ResourceName, resource)
+	for k, v := range tp.meta {
+		span.SetTag(k, v)
+	}
+	if meta, ok := ctx.Value(spanTagsKey).(map[string]string); ok {
+		for k, v := range meta {
+			span.SetTag(k, v)
+		}
+	}
+
+	if sqlCommentCarrier != nil && tp.cfg.commentInjectionMode != commentInjectionDisabled {
+		injectionOpts := injectionOptionsForMode(tp.cfg.commentInjectionMode, sqlCommentCarrier.DiscardDynamicTags)
+		err := tracer.InjectWithOptions(span.Context(), sqlCommentCarrier, injectionOpts...)
+		if err != nil {
+			// this should never happen
+			fmt.Fprintf(os.Stderr, "contrib/database/sql: failed to inject query comments: %v\n", err)
+		}
+	}
+
+	return span
+}
+
 // tryTrace will create a span using the given arguments, but will act as a no-op when err is driver.ErrSkip.
 func (tp *traceParams) tryTrace(ctx context.Context, qtype queryType, query string, startTime time.Time, err error) {
 	if err == driver.ErrSkip {
```

Comment:
The new feature is disabled by default and the intent is to not break things for someone running with the disabled feature. However, the nature of how spans were created in one go _after_ a database operation isn't compatible with the new feature which requires the span to be created prior to that operation so that we can inject trace id, span id and sampling priority as a sql comment on queries that get executed. That pattern is usually fine and is pretty standard except that here (on this specific line), we have this logic that skips the span creation if the driver returns `driver.ErrSkip`. We can't have that logic in the `tryStartTrace` because we execute it before the operation runs. 

I'm curious if anyone has any ideas on how best to do this. I could have a hard fork to check if the comment injection mode is disabled and just use the old pattern with `tryTrace` instead of `tryStartTrace` but we'd still have this potential problem when comment injection is enabled. 

cc @gbbr since I think you worked on that issue in the past.

Reply:
Interesting question, I'm not sure of the final answer but here's some thoughts:
1) I think the case where `tryTrace` sees a driver return `ErrSkip` should be rare since the re-write of this areas in this PR #1167. It seems most drivers that implement sql optional interfaces like `ExecerContext` generally won't return a `driver.ErrSkip`(as they actually have implemented the interface). I did a quick search of pgx, pq, sqlx and mysql, and only the mysql driver ever returns it. Given that it _can_ happen it's probably better to explicitly handle this case to avoid the rare bug, but it's worth knowing it's rare in case we can find a solution that covers us but might be an expensive operation.
2) Perhaps we could drop or untrack a span in this situation? I'm not sure if there's any other integrations that do this that we could steal that functionality from. An alternate way to think of this might be to "prepare" but not _start_ the span? That way the trace_id and span_id are available for the sqlcommenter, but if the driver has to back out with an ErrSkip the span is never started (and can be tracked by the appropriate function when the sql package does its fallback logic)*

*I could see this approach rapidly getting complex as things like "long running span tracking" are added which might notify the backend about open spans.

Reply:
> Perhaps we could drop or untrack a span in this situation? I'm not sure if there's any other integrations that do this that we could steal that functionality from. An alternate way to think of this might be to "prepare" but not start the span?

I thought about this but wasn't sure how much it fits with existing patterns. I'm not seeing it but is there a way right now to create a span without starting it?

Reply:
> I thought about this but wasn't sure how much it fits with existing patterns. I'm not seeing it but is there a way right now to create a span without starting it?

I don't see a way to do that currently via the public API. Hypothetically we might be able to _add_ it somehow but I'm not sure that's the best approach since this seems like a very specific need.

I have another idea. We could still create / start the span *after* the SQL operations. (correct me if i'm missing one here) but I think the only necessary field that we don't have for the sql comment generation is the span_id. The trace_id is already known without starting a new span. What if we generated a span_id upfront then if we end up actually sending the comment (and don't get `ErrSkip`) we can use `tracer.WithSpanID` to ensure the IDs line-up